### ISSUE TITLE
Deleting custom field populated with data results in phantom field

### ISSUE THREAD

**Issue by waydh** [nautobot/nautobot#739]
### Environment
* Python version: 3.6.8
* Nautobot version: 1.1.0

### Steps to Reproduce
1. Add a custom field to a dcim/device page
2. Add data to the field (via GUI or API does not matter)
3. Go to Admin page and delete the custom field
4. Return to the device page and edit something
5. Click Update and receive error, "Unknown field name '<FIELDNAME>'  in custom field data"
6. Go to the Admin page and re-create the custom field
7. Return to the device and delete the data from the custom field and Update
8. Go to the Admin page and delete the custom field
9. Now the error will not occur
10. If you do this with two fields (field1 and field2 eg) and put data in both, then delete both from Admin page, you will be in a constant loop of errors (Unknown field 1 then after fixing field1, Unknown field2) . Fix field 1, but you can't update as field2 is in limbo, fix field2, can't update as field 1 is in limbo
11. It seems that the delete custom field on the Admin page is not deleting the field data and not detaching the field from any device it had data in

### Expected Behavior
Deleting custom field from Admin page completely erases the custom field and its data so that no error occurs

### Observed Behavior
"Unknown field name '<FIELDNAME>'  in custom field data

See above steps for more information.

**Comment by waydh** [comment]
I just tried again to remove a custom field, even after clearing the data in the field and then deleting it, I am still getting the Error in Unknown field name "blah" . I can't remove custom fields at all.

I'm fairly certain this was not an issue in 1.0.3.

**Comment by glennmatthews** [comment]
Thanks for the report!

Can you confirm that with 1.1.0 you have the Celery worker (`nautobot-server celery worker`) running in your environment? If so, can you also confirm whether the issue persists if you wait, say, 30 seconds between step 3 and step 4?

I ask because the cleanup after deleting a custom field is handled as a background task, so it's possible that either the task isn't running at all (if the Celery worker isn't running), or that you're hitting a timing/race condition (which would still be considered a bug, to be clear) where you're updating a device before it's been properly cleaned by the background task.

Also, in 1.1.0 there is a new management command, `nautobot-server fix-custom-fields`, which you might try running as a last resort to see if it resolves the issue for you - obviously this isn't the preferred solution, but if it unblocks you in the meantime I wanted you to at least be aware of it. 

Thanks again!

**Comment by jathanism** [comment]
> Thanks for the report!
> 
> Can you confirm that with 1.1.0 you have the Celery worker (`nautobot-server celery worker`) running in your environment? If so, can you also confirm whether the issue persists if you wait, say, 30 seconds between step 3 and step 4?
> 
> I ask because the cleanup after deleting a custom field is handled as a background task, so it's possible that either the task isn't running at all (if the Celery worker isn't running), or that you're hitting a timing/race condition (which would still be considered a bug, to be clear) where you're updating a device before it's been properly cleaned by the background task.
> 
> Also, in 1.1.0 there is a new management command, `nautobot-server fix-custom-fields`, which you might try running as a last resort to see if it resolves the issue for you - obviously this isn't the preferred solution, but if it unblocks you in the meantime I wanted you to at least be aware of it.
> 
> Thanks again!

@waydh For reference: https://nautobot.readthedocs.io/en/stable/release-notes/version-1.1/#background-tasks-now-use-celery-223

**Comment by waydh** [comment]
Hello all,

Thank you for your responses.

Running nautobot-server celery worker shows celery is running, but then a tonne of errors spew past - I could not catch them all - I have attached what I could get. It seemed to freeze then and I had to break out. Now running the command shows just the celery running out put. The field issue remained.

Nautobot-server fix_custom_fields also produced errors - the field issue is still there. Interestingly, fix_custom_fields is complaining about a field I added after the one that broke.

The original field I created is called serial (I hadn't noticed serial field was already available on device page...) this is the one I deleted and is causing the error - Unknown field etc.

The field fix_custom_fields is complaining about is one I added after serial, it's called architecture.

I've attached the output as text files.

Cheers!
[celery errors on first run.txt](https://github.com/nautobot/nautobot/files/6889451/celery.errors.on.first.run.txt)
[celery output on second run.txt](https://github.com/nautobot/nautobot/files/6889452/celery.output.on.second.run.txt)
[fix_custom_fields output.txt](https://github.com/nautobot/nautobot/files/6889453/fix_custom_fields.output.txt)

**Comment by glennmatthews** [comment]
Following up on this - the Celery errors on the first run are due to #744, which should be resolved in the next patch release. (You can try patching in #763 in the short term and see if it helps.) 

Looks like there's some additional error handling needed in `fix_custom_fields` as it shouldn't be erroring out in that way.

**Comment by waydh** [comment]
Great! Thanks for working on this. I'll hold off rolling back to 1.0.3 and wait for the next release.

Cheers!

**Comment by waydh** [comment]
I've upgraded to 1.1.1 and am still having the exact same issue.

I've deleted the custom field and run nautobot-server celery worker and get the following output;-

 -------------- celery@hidden v5.1.2 (sun-harmonics)
--- ***** ----- 
-- ******* ---- Linux-4.18.0-305.3.1.el8.x86_64-x86_64-with-centos-8.4.2105 2021-08-06 01:05:52
- *** --- * --- 
- ** ---------- [config]
- ** ---------- .> app:         nautobot:0x7fa47ca7ad68
- ** ---------- .> transport:   redis://localhost:6379/0
- ** ---------- .> results:     redis://localhost:6379/0
- *** --- * --- .> concurrency: 20 (prefork)
-- ******* ---- .> task events: OFF (enable -E to monitor tasks in this worker)
--- ***** ----- 
 -------------- [queues]
                .> celery           exchange=celery(direct) key=celery
                

[2021-08-06 01:05:55,665: ERROR/ForkPoolWorker-14] Custom field with ID dd5b62b7-803c-405b-85d0-331832424910 not found, failing to provision.

It still can't do the work to remove the custom field correctly. No updates can be performed on devices unless I recreate that field.

nautobot-server fix_custom_fields is still complaining about the custom field that was added after the one that was deleted. 'serial' was the custom field that was deleted, 'architecture' is a new field that was added after 'serial'. 

Processing ContentType dcim | device
Traceback (most recent call last):
  File "/opt/nautobot/bin/nautobot-server", line 8, in <module>
    sys.exit(main())
  File "/opt/nautobot/lib64/python3.6/site-packages/nautobot/core/cli.py", line 62, in main
    initializer=_configure_settings,  # Called after defaults
  File "/opt/nautobot/lib64/python3.6/site-packages/nautobot/core/runner/runner.py", line 266, in run_app
    management.execute_from_command_line([runner_name, command] + command_args)
  File "/opt/nautobot/lib64/python3.6/site-packages/django/core/management/__init__.py", line 401, in execute_from_command_line
    utility.execute()
  File "/opt/nautobot/lib64/python3.6/site-packages/django/core/management/__init__.py", line 395, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/opt/nautobot/lib64/python3.6/site-packages/django/core/management/base.py", line 330, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/opt/nautobot/lib64/python3.6/site-packages/django/core/management/base.py", line 371, in execute
    output = self.handle(*args, **options)
  File "/opt/nautobot/lib64/python3.6/site-packages/nautobot/extras/management/commands/fix_custom_fields.py", line 26, in handle
    if not obj._custom_field_data[custom_field.name]:
KeyError: 'architecture'


Thanks!

**Comment by waydh** [comment]
Ok, further experimentation.

Deleting the 'architecture' custom field, then running nautobot-server celery-worker produced no errors. Then running nautobot-server fix_custom_fields worked - it added missing custom fields to hundreds of devices. I was able to update device details again. 

I then added the custom field 'architecture' back, and was able to update device details ok. Running nautobot-server fix_custom_fields failed on the 'architecture' field again, running nautobot-server celery worker, and then nautobot-server fix_custom_fields worked... and it added all the custom fields to hundreds of devices - it looks like the same ones.

This brings up another issue, it'd be really nice to have two device types - network and compute (like how there is device and virtual machine now). In HPC we have 1000s of compute devices or more, as well as hundreds and more network devices - we need to have different custom fields for both of these and a way to mark them as network or compute.

I might log a feature request for this bit.

Thanks for all your help!
